libkbfs/bare_root_metadata.go

```diff
@@ -510,70 +516,75 @@ func (md *BareRootMetadata) getTLFKeyBundles(keyGen KeyGen) (
-	if u, ok1 := wkb.WKeys[user]; ok1 {
-		info, ok := u[key]
-		return info, ok, nil
-	} else if u, ok1 = rkb.RKeys[user]; ok1 {
-		info, ok := u[key]
-		return info, ok, nil
-	}
-	return TLFCryptKeyInfo{}, false, nil
-}
-
-// GetTLFCryptPublicKeys returns the public crypt keys for the given user
-// at the given key generation.
-func (md *BareRootMetadata) GetTLFCryptPublicKeys(
-	keyGen KeyGen, user keybase1.UID) (
-	[]keybase1.KID, bool) {
-	wkb, rkb, err := md.getTLFKeyBundles(keyGen)
-	if err != nil {
-		return nil, false
+		return false
 	}
 
-	if u, ok1 := wkb.WKeys[user]; ok1 {
-		return u.GetKIDs(), true
-	} else if u, ok1 = rkb.RKeys[user]; ok1 {
-		return u.GetKIDs(), true
-	}
-	return nil, false
+	return (len(wkb.WKeys[user]) > 0) || (len(rkb.RKeys[user]) > 0)
 }
 
-// GetTLFEphemeralPublicKey returns the ephemeral public key used for
-// the TLFCryptKeyInfo for the given user and device.
-func (md *BareRootMetadata) GetTLFEphemeralPublicKey(
-	keyGen KeyGen, user keybase1.UID,
-	currentCryptPublicKey CryptPublicKey) (TLFEphemeralPublicKey, error) {
+// GetTLFCryptKeyInfo returns all the necessary info to construct the
+// TLF crypt key for the given key generation, user, and device
+// (identified by its crypt public key), or false if not found. This
+// returns an error if the TLF is public.
+func (md *BareRootMetadata) GetTLFCryptKeyInfo(
+	keyGen KeyGen, user keybase1.UID, key CryptPublicKey) (
+	TLFEphemeralPublicKey, EncryptedTLFCryptKeyClientHalf,
+	TLFCryptKeyServerHalfID, bool, error) {
 	wkb, rkb, err := md.getTLFKeyBundles(keyGen)
 	if err != nil {
-		return TLFEphemeralPublicKey{}, err
+		return TLFEphemeralPublicKey{},
+			EncryptedTLFCryptKeyClientHalf{},
+			TLFCryptKeyServerHalfID{}, false, err
 	}
 
-	info, ok, err := md.GetTLFCryptKeyInfo(
-		keyGen, user, currentCryptPublicKey)
-	if err != nil {
-		return TLFEphemeralPublicKey{}, err
+	dkim := wkb.WKeys[user]
+	if dkim == nil {
+		dkim = rkb.RKeys[user]
+		if dkim == nil {
+			return TLFEphemeralPublicKey{},
+				EncryptedTLFCryptKeyClientHalf{},
+				TLFCryptKeyServerHalfID{}, false, nil
+		}
 	}
+	info, ok := dkim[key.kid]
 	if !ok {
 		return TLFEphemeralPublicKey{},
-			TLFEphemeralPublicKeyNotFoundError{
-				user, currentCryptPublicKey.kid}
+			EncryptedTLFCryptKeyClientHalf{},
+			TLFCryptKeyServerHalfID{}, false, nil
 	}
 
 	if info.EPubKeyIndex < 0 {
-		return rkb.TLFReaderEphemeralPublicKeys[-1-info.EPubKeyIndex], nil
+		index := -1 - info.EPubKeyIndex
+		keyCount := len(rkb.TLFReaderEphemeralPublicKeys)
+		if index >= keyCount {
+			return TLFEphemeralPublicKey{},
+				EncryptedTLFCryptKeyClientHalf{},
+				TLFCryptKeyServerHalfID{}, false,
+				fmt.Errorf("Invalid reader key index %d >= %d",
+					index, keyCount)
+		}
+		return rkb.TLFReaderEphemeralPublicKeys[index],
+			info.ClientHalf, info.ServerHalfID, true, nil
+	}
+
+	index := info.EPubKeyIndex
+	keyCount := len(wkb.TLFEphemeralPublicKeys)
+	if index >= keyCount {
```

Comment:
would be nice to DRY this check with the above one, but I don't think it's a big deal